net: option to disallow v1 connection on ipv4 and ipv6 peers

stratospher opened the pull request:
resolves #29618.

This PR adds a config flag option `-v2onlyclearnet` which disallows outbound v1 connections on ipv4 and ipv6 networks since they're unencrypted. outbound v1 connections are still possible from tor/i2p/cjdns peers since they're encrypted networks.

- v1 outbound connections to peers are not attempted
- v1 reconnections are not attempted (if peer is falsely advertised as v2 peer and we attempt a v2 connection but it fails, we do not attempt a v1 reconnection when `-v2onlyclearnet` is switched on)

Currently 70% of reachable nodes in network supporting v2 (according to https://21.ninja/reachable-nodes/node-service-share/ and https://bitnodes.io/nodes). Also found 81% of addresses supporting v2 in the addrman of nodes I checked - personal node and the nodes in [peer.observer](https://peer.observer/addrman/alice). (You can use [this branch ](https://github.com/stratospher/bitcoin/commits/addrinfo_service_display/)to check the % of v2 addresses in a node's addrman.)

stratospher committed:
init: add -v2onlyclearnet config option

if this option is set by the user, v1 connections on unencrypted
networks like IPV4/IPV6 will be disallowed. Only users with real
need are recommended to turn this on because it could risk network
partitioning in the unlikely scenario that everyone turns it on.

stratospher committed:
net: disable v1 connections, reconnections on clearnet

if `-v2onlyclearnet` is turned on,
- v1 addresses from addrman are dropped if selected and manual
connections aren't attempted for outbound connections if it's
from IPV4/IPV6 networks.
- v1 downgrade mechanism is not attempted if v2 connection wasn't
successful

stratospher committed:
net: add option in CConman to allow only v2 clearnet connections

a boolean option `m_v2only_clearnet` is introduced in CConman
which will (in a later commit) store if the user wishes to
establish only outbound v2 connections to IPv4/IPv6 peers
since they are unencrypted.

this option is accessible outside CConman using
`RequiresV2ForOutbound()` function with the IP address we're trying
to connect to passed as an argument.

stratospher committed:
test: Check that v1 connections to clearnet peers don't work

when `-v2onlyclearnet` is turned on:
- v1 connections to clearnet peers don't work
- v2 connections to clearnet peers work
- v1 connections to tor/i2p/cjdns peer works

a proxy is used because otherwise NET_UNROUTABLE is the default
network in the tests.

mzumsande commented:
Concept ACK

> In #29618 and here, I don't see the motivation to forbid v1 connections.

Well, to gain the benefits of BIP324. For example if you are concerned about someone inspecting the traffic to your node it really doesn't matter how many BIP324 peers you have if you have just one unencrypted connection. You could also run tor-only in this case, but that has its own disadvantages and `-onlynet=onion` shares the exact same downside that if everyone did it, the network would split. If some, but not all nodes who are running only on privacy networks would be willing to also have clearnet peers with this option, that would be a good thing in my opinion.

>  if V2-only is widespread

that would be a problem, but `-onlynet`, `-blocksonly` etc. have that exact same problem. I see this as an option you would only choose if you actually need it because you are concerned about unencrypted traffic.

jonatack commented:
> You could also run tor-only in this case, but that has its own disadvantages and `-onlynet=onion` shares the exact same downside that if everyone did it, the network would split. 

Our docs address this to an extent [here](https://github.com/bitcoin/bitcoin/blob/master/doc/i2p.md#additional-configuration-options-related-to-i2p) and [here](https://github.com/bitcoin/bitcoin/blob/master/doc/cjdns.md#additional-configuration-options-related-to-cjdns). As those docs point out, `-onlynet` only affects outbound connections (not inbound and manual ones), and we suggest that it be used with multiple networks.  My blog goes further into it [here](https://jonatack.github.io/articles/using-alternative-p2p-networks-with-bitcoin-core#why-use-multiple-bitcoin-p2p-networks) and [here](https://jonatack.github.io/articles/using-alternative-p2p-networks-with-bitcoin-core#considerations-with-running-tor-only).

> If some, but not all nodes who are running only on privacy networks would be willing to also have clearnet peers with this option, that would be a good thing in my opinion.

Agree. My node, for instance, runs over tor/i2p/cjdns and makes manual connections to trusted clearnet peers.

> if V2-only is widespread, it may be hard to eclipse V2-only node, but then it becomes easy to eclipse a V1 node which has a problem of finding peers to connect to, so an attacker starts a lot of nodes that do accept V1 connections.

Agree. IIUC users will probably increasingly need to accept v2, whether they wish to or not. I'm unaware if there are downsides to that. We generally try to avoid forcing users to do things, but the torv2->torv3 migration a few years ago was one example (bitcoin-over-tor users had to upgrade to a recent version of bitcoin core); perhaps v2 p2p is another. I suppose my concept ack on this would depend on that.

vasild commented:
> For example if you are concerned about someone inspecting the traffic to your node it really doesn't matter how many BIP324 peers you have if you have just one unencrypted connection.

True. Lets dive a bit deeper - what are the reasons to be concerned about somebody inspecting the traffic to my node?

jonatack commented on src/init.cpp:
```diff
@@ -550,6 +550,7 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
     argsman.AddArg("-i2pacceptincoming", strprintf("Whether to accept inbound I2P connections (default: %i). Ignored if -i2psam is not set. Listening for inbound I2P connections is done through the SAM proxy, not by binding to a local address and port.", DEFAULT_I2P_ACCEPT_INCOMING), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-onlynet=<net>", "Make automatic outbound connections only to network <net> (" + Join(GetNetworkNames(), ", ") + "). Inbound and manual connections are not affected by this option. It can be specified multiple times to allow multiple networks.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-v2transport", strprintf("Support v2 transport (default: %u)", DEFAULT_V2_TRANSPORT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-v2onlyclearnet", strprintf("Disallow v1 connections on IPV4/IPV6 (default: %u). Enabling this is not recommended unless absolutely necessary, as it may risk network partitions if all users enable it.", false), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
I might be too focused on the risks rather than benefits, but this is unfortunately only one brigading campaign by one or more prominent bitcoiners on social media away from being potentially adopted by many nodes.

With -onlynet, the network partition risk is in practice opt-in, i.e. chosen by a minority subset for themselves.  Whereas here, the network partition risk could be imposed on a future minority of users that are running earlier versions of Bitcoin Core. This seems at odds with maintaining backward compatibility for that minority.

jonatack commented on src/net.h:
```diff
@@ -1572,6 +1577,13 @@ class CConnman
+    /**
+     * option for disabling v1 connections on IPV4 and IPV6.
+     * connections on IPV4/IPV6 need to be v2 connections.
+     * connections on Tor/I2P/CJDNS maybe v1 or v2 connections.
```
```suggestion
     * connections on Tor/I2P/CJDNS can be v1 or v2 connections.
```

jonatack commented on src/net.cpp:
```diff
@@ -2943,6 +2953,25 @@ void CConnman::OpenNetworkConnection(const CAddress& addrConnect, bool fCountFai
+        SplitHostPort(std::string(pszDest), port, host);
+        if (LookupHost(host, false).has_value()) {
+            netclass = LookupHost(host, false).value().GetNetClass();
+        } else if (LookupHost(host, true).has_value()) {
```
```suggestion
        } else if {
            (LookupHost(host, true).has_value()) {
```

mzumsande commented:
> True. Lets dive a bit deeper - what are the reasons to be concerned about somebody inspecting the traffic to my node?

Some thoughts off the top my head:
- You don't want others (e.g. network admins, internet providers, etc.) to know that you are running a bitcoin node at all. Maybe it's not allowed in your network / country etc., maybe for other reasons such as personal security.
- Transaction privacy

In both of these cases, if you accept inbounds they could just connect to you as an alternative (but they would need to have a suspicion, whereas detection of bitcoin-related network data could be automated). So another alternative would be to only disallow v1 connections for outbound peers. In that case, if you don't want any v1 connections at all you'd have to run with `-nolisten`.

vasild commented:
> > what are the reasons to be concerned about somebody inspecting the traffic to my node?
> 
> * You don't want others (e.g. network admins, internet providers, etc.) to know that you are running a bitcoin node at all. Maybe it's not allowed in your network / country etc., maybe for other reasons such as personal security.

This is what is worrying me - V2-only will not hide the fact that I am running a bitcoin node, but may give the false impression that it does. This is dangerous. My node will connect to publicly known bitcoin nodes and the mere fact that I have a TCP connection to an `addr:port`, where it is known that a bitcoin node is running, is already revealing enough. No need to inspect the traffic itself.

Another possible misunderstanding is that V2-only makes the traffic impossible to spy. I can imagine a blatant MITM, which is detectable because the session keys would differ, should somebody bother to check them. Once I detect this, whom am I going to complain to? My oppressive gov/ISP/whatever, same one that is doing the spying? Or the ISP can outright redirect my connections to `addr:port` to their spy node at `spyaddr:port` (I think I am talking to `addr`, but `addr` does not even have a connection from me). The point is that the traffic is spyable. Then, what is it so much that I want to keep away from spies? The data being transferred is all public, except if we are talking about me broadcasting my own transactions, there you go:

> * Transaction privacy

V2-only may give the false impression that it breaks the link between transaction origin and IP address / geolocation. Even with unspyable p2p encryption, I may have a connection to a spy bitcoin node which then sees my traffic. Like you said "it really doesn't matter how many BIP324 peers you have if you have just one unencrypted connection" I think this is the same as "it does not matter even if all your connections are encrypted, if you have just one (encrypted) connection to a spy node".

stratospher commented:
@vasild V2 only doesn't protect against active attacks like the ones you've mentioned which require resources from the other party to run a node to spy the traffic.

V2 only protects against a cheaper attack vectors possible because of passive inspection of network traffic flowing through different medium. They don't need the other party to run a node to spy.

> > True. Lets dive a bit deeper - what are the reasons to be concerned about somebody inspecting the traffic to my node?

> Some thoughts off the top my head:
> - You don't want others (e.g. network admins, internet providers, etc.) to know that you are running a bitcoin node at all. Maybe it's not allowed in your network / country etc., maybe for other reasons such as personal security.
> - Transaction privacy

Yes, for the same reason as wanting to encrypt network traffic in the first place and this just offers a stronger guarantee when running a node with v2 support that unencrypted clearnet bitcoin traffic is not being collected and sold by other parties. Deep packet inspection, keyword filtering are cheaper for ISPs/firewalls to pull off and v2 only would protect against this.

vasild commented:
A V2-only option will:

* not hide the fact that one runs bitcoin node (that is true regardless of encryption) and
* not stop even moderately motivated actor from spying for transaction origin (either by opening a P2P connection to the target or bricking V2 encryption or outright redirecting connections to their node)

but it will give the false impression that it does those two things. Further, it will make it more difficult for old nodes to find peers to connect to.

Concept NACK because of that.

dergoegge commented:
Concept ACK

This option seems useful for users that want to:

* prevent passive spying on their connections
* increase the cost of tampering with the data exchanged on their connections
* increase the cost of detecting that they run a bitcoin node

> Currently 59.71% of network supports v2

Have you checked how diverse these ~60% of v2 supporting nodes are? An extreme example: if they are all on running on AWS then I don't think we should add the option at this time.

vasild commented:
> increase the cost of detecting that they run a bitcoin node

How? Given that the spy does not need to inspect the traffic at all:

> My node will connect to publicly known bitcoin nodes and the mere fact that I have a TCP connection to an addr:port, where it is known that a bitcoin node is running, is already revealing enough.

jonatack commented:
> However, as I pointed out in the linked issue, I am somewhat concerned about a potential future where it becomes harder for other/older software to connect to the network if large swaths move to be v2-only. Because of that, I wonder if it isn't better to make this only apply to outbound connections, and advise people who want more private operation to not listen for inbound connections.

Yes. I may be potentially concept ack here if this proposal adopts the suggestions in https://github.com/bitcoin/bitcoin/issues/29618#issuecomment-1994415042 and https://github.com/bitcoin/bitcoin/issues/29618#issuecomment-2384291593. (Along with perhaps additional user documentation for node runners, maybe in the relevant config option helps and/or in `/doc`.)

jonatack commented:
> However, as I pointed out in the linked issue, I am somewhat concerned about a potential future where it becomes harder for other/older software to connect to the network if large swaths move to be v2-only. Because of that, I wonder if it isn't better to make this only apply to outbound connections, and advise people who want more private operation to not listen for inbound connections.

Yes. I may be potentially concept/approach ack here if this proposal adopts the suggestion above / in https://github.com/bitcoin/bitcoin/issues/29618#issuecomment-1994415042 or https://github.com/bitcoin/bitcoin/issues/29618#issuecomment-2384291593. (Along with perhaps additional user documentation for node runners, maybe in the relevant config option helps and/or in `/doc`.)

brunoerg commented:
A mutation testing report for this PR is available at: https://brunoerg.xyz/mutation-core-front

laanwj commented:
> Well, to gain the benefits of BIP324. For example if you are concerned about someone inspecting the traffic to your node it really doesn't matter how many BIP324 peers you have if you have just one unencrypted connection. 

Right. Would it maybe make sense to have an option to restrict transaction broadcast an relay to v2-only, for IPv4 and IPv6 connections? Instead of not connect at all? Eg consider v1 connections blocks-only (except those over Tor/I2P). This alleviates the risk of network partition of consensus. While still providing a bit of added privacy.

stratospher commented:
I've rebased and changed the approach in the PR to have v2 only outbound connections. So there's null partition risk because of the option now.

> Right. Would it maybe make sense to have an option to restrict transaction broadcast and relay to v2-only, for IPv4 and IPv6 connections? Instead of not connect at all? Eg consider v1 connections blocks-only (except those over Tor/I2P). This alleviates the risk of network partition of consensus. While still providing a bit of added privacy for where it matters.


I don't think just sending transactions over encrypted connections gives that much extra privacy.

Passive attackers can still:
1. _cheaply_ tell you're running a bitcoin node looking at the network magic bytes.
2. there's transaction lengths which can be used to distinguish 1 transaction from another.

Having all network traffic as encrypted would give better privacy when dealing with 1.

laanwj commented:
> I don't think just sending transactions over encrypted connections gives that much extra privacy.

i'm not advicating to send *just transactions*, v2 connections would have blocks+transactions, all of it. Not sending transactions over v1 makes sure that transactions aren't sent over the network in plaintext. 

In the long term, if the idea is to phase out P2P v1, i think demotiing it to a blocksonly-protocol makes sense. It holds the network together against eclipse attacks and accidental partitioning, but doesn't reveal anything privacy sensitive.

Sure, passive attackers can still see packet sizes in the v2 connections, but that doesn't actually change if you refuse to accept and make v1 connections at all?

> cheaply tell you're running a bitcoin node looking at the network magic bytes.

i still don't believe there is any hiding like that. There's so many tells for passive attackers: using the DNS seeds, connecting to certain nodes, using port 8333. Sure, if you manage to avoid all of those, it's less apparent, but that's so much work and easy to screw up with a small mistake.

vasild commented:
Before I could see the purpose of the full v2only option: 1. to hide the fact that a bitcoin node is running and 2. hide its own transactions. However I think it would achieve neither of those, but I could see how it looked like it would. Now it is changed:

> I've rebased and changed the approach in the PR to have v2 only outbound connections

Does it still have the same motivation: 1. and 2.? I think the idea in https://github.com/bitcoin/bitcoin/issues/29618 is that v1 connections are somehow harmful and we must not have any of them. Now with this PR we would allow v1 inbound connections? @iotamega, do you think this would resolve #29618? Once the node has e.g. 30 connections and some of them are v1 and some v2, I think it is irrelevant which are inbound and which are outbound.

Consider outbound v1 connections to clearnet but via the Tor proxy (via the Tor network and Tor exit nodes), if `-proxy=` is given. Assuming there is some reason to avoid v1 outbound (which I don't see), should those be avoided as well?

1440000bytes reviewed (changes requested):
NACK

This pull request and related efforts are bad for bitcoin in lot of ways not just partition risk.

We need to fix other things before trying to achieve this. Disappointed that some devs keep repeating it will hide that you are running a node.

luke-jr reviewed (changes requested):
No opinion on concept. Just code review.

luke-jr commented on src/init.cpp:
```diff
@@ -1843,6 +1846,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     connOptions.m_peer_connect_timeout = peer_connect_timeout;
     connOptions.whitelist_forcerelay = args.GetBoolArg("-whitelistforcerelay", DEFAULT_WHITELISTFORCERELAY);
     connOptions.whitelist_relay = args.GetBoolArg("-whitelistrelay", DEFAULT_WHITELISTRELAY);
+    connOptions.disable_v1conn_clearnet = args.GetBoolArg("-v2transport", DEFAULT_V2_TRANSPORT) && args.GetBoolArg("-v2onlyclearnet", false);
```
```suggestion
    connOptions.disable_v1conn_clearnet = args.GetBoolArg("-v2onlyclearnet", false);
```

The InitError above means this should be safe

ajtowns reviewed (commented):
Concept review:

I think this has two potential benefits:

 * avoids cleartext relay of transactions, particularly your own. #29415 is better at solving that problem, but a small step is also a good thing. this is only useful if you also disable inboud clearnet connections however.
 * avoids easy detection that you're running a bitcoin node by stateless packet connection. if you're a listening node, then you still advertise your ip over the bitcoin network, so it's not hard to find you, in that case. traffic analysis or similar also likely gives this away, but that is significantly more work.

Both these suggest to me that this option is only really meaningful for non-listening nodes. Perhaps either automatically setting nolisten, or adding a "only listen on non-clearnet networks" option and automatically setting that would be good.

Only allowing v2 clearnet inbounds might make sense, but it seems concerning for eclipsing old nodes that don't support v2. Downgrading clearnet v1 inbounds to blocksonly mode might be an acceptable compromise there. Probably old nodes would still want some way to broadcast their transactions to the network, however, which might be hard to support in that model.

Particularly if this option is only used by nodes that aren't listening on clearnet anyway, I think this is a reasonable option to have.

Approach ACK 45bd8914658a675d00aa9c83373d6903a8a9ece8

mzumsande reviewed (commented):
A year later, adoption is now almost at ~70% according to https://21.ninja/reachable-nodes/node-service-share/ - could update the OP with that info.

stratospher committed:
net: add helper to identify clearnet networks

only-clearnet condition is used in a few other places as well.

fjahr reviewed (commented):
Concept ACK

I agree with @ajtowns that coupling this with nolisten makes sense and also means that this should not be too dangerous. Rather than turning that on by default and potentially risking the user don't fully understand the implications of this, the option could also error if nolisten isn't set.

sipa commented:
Concept ACK. I think deployment of v2-capable nodes is sufficient that it's reasonable to offer an outbound-v2-only option.

mzumsande reviewed (commented):
> deny v1 incoming (only allow v2 incoming) which is a problem if this is widely adopted because then old nodes that do not support v2 will have a hard time finding peers to connect to and will be subject to eclipse from an attacker who runs many v1-accepting nodes.

I still don't find the reason for tying it to `-listen=0` instead of also disallowing v1 inbounds compelling. Yes, if everyone disabled v1 incoming peers that could lead to v1 nodes being eclipsed. But with the suggested approach, if too many nodes just disabled incoming connections entirely to be able to use the option, that wouldn't be any better, because now _every_ node would have trouble finding peers / risk being eclipsed. 

If the downsides are documented well in the helptext, I don't see a difference to many other existing optional settings that would be bad if everyone used them. (`-listen`, `-blocksonly`, `-onlynet` etc.). But just comparing the two options in a vacuum, allowing only v2 inbounds seems better than allowing no inbounds at all.

vasild commented:
From the [issue](https://github.com/bitcoin/bitcoin/issues/29618#issuecomment-2388394165) that prompted this:
> I asked for this because of a friend in a country where they are having issues, very real issues and they did not want to doxx themselves

I still think that this is misguided and can get people into "very real issues" if such an option gives the impression that it:

1. Hides the fact that you run a bitcoin node. It does not at all because you will still connect to others at port `8333`, and that is the easiest way to detect bitcoin traffic, no need for spy nodes or inspecting the traffic contents.

2. Hides you as originator of your own transactions from your ISP. A little bit it does, but not much as commented in details above. I [think](https://github.com/bitcoin/bitcoin/pull/30951#issuecomment-3497705768) that all-v2-connections is the same as at-least-one-v2-connection for concealing the transaction origin.

Anyway, if it is made clear that a v2only option does not achieve 1. or 2., then I guess it is fine. V2 adoption has increased and there are enough knowledgeable people commenting here, so with this post I would like to withdraw my "Concept NACK" and leave it in your capable hands.

Concept 0

<!--meta-tag:bot-skip-->

stratospher commented on src/net.cpp:
```diff
@@ -2498,6 +2498,11 @@ bool CConnman::MultipleManualOrFullOutboundConns(Network net) const
     return m_network_conn_counts[net] > 1;
 }
 
+bool CConnman::DisableV1OnClearnet(Network net) const
```
makes sense. changed it to `RequiresV2ForOutbound`. hopefully it's better.

stratospher commented:
Thank you all for the very helpful feedback on this! and sorry for getting late with the update.

I've pushed an [update](https://github.com/bitcoin/bitcoin/compare/27e90008835a4c980447d97e80910742e041dfaf..14ee29dedc36611f9d33a207b7c5b25a7c9b89ea) addressing all the review comments except @mzumsande's conceptual concern about whether
`v2onlyclearnet` should accept v2 inbounds as well so that the network doesn't face a shortage of listening capacity in the worst case scenario where a huge fraction of network turns it on.

Quick summary of the [update](https://github.com/bitcoin/bitcoin/compare/27e90008835a4c980447d97e80910742e041dfaf..14ee29dedc36611f9d33a207b7c5b25a7c9b89ea):
- More description in help text to highlight that `-v2onlyclearnet` doesn't hide the fact that you're running a bitcoin node. (thanks @vasild!)
- `-v2onlyclearnet`, `-listen` arg interaction which errors if we use `-v2onlyclearnet` on a listening node (thanks @ajtowns!)
-  naming consistency in functions/variables to use "v2 only" language instead of mixing it with potentially confusing "disable v1" language also. Ex: s/`DisableV10nClearnet`/`RequiresV2ForOutbound` (thanks @fjahr!)
- introduced `DEFAULT_DISABLE_V1CONN_CLEARNET=false` (thanks @laanwj!)

regarding https://github.com/bitcoin/bitcoin/pull/30951#pullrequestreview-3429992454:
> I still don't find the reason for tying it to -listen=0 instead of also disallowing v1 inbounds compelling. Yes, if everyone disabled v1 incoming peers that could lead to v1 nodes being eclipsed. But with the suggested approach, if too many nodes just disabled incoming connections entirely to be able to use the option, that wouldn't be any better, because now every node would have trouble finding peers / risk being eclipsed.


I assumed people who didn't want network traffic to be spyed on are already people who didn't accept (possibly spy) connections from others (people running private nodes) and didn't think of the impact on the listening slots in the network. This doesn't necessarily need to be the case though. `-v2onlyclearnet` being used with `listen=0` would also need `-listenonion=0`, so we'd also be turning off tor/i2p inbounds.

However this feels like lots of users setting listen=0 problem, and we're showing someone a reason for setting listen=0 + making that future slightly more probable. Will need to think more about this and curious about other people's thoughts on https://github.com/bitcoin/bitcoin/pull/30951#pullrequestreview-3429992454 as well.

fjahr commented on src/netbase.cpp:
```diff
@@ -141,6 +141,8 @@ std::vector<std::string> GetNetworkNames(bool append_unroutable)
     return names;
 }
 
+bool isClearnet(const enum Network net) { return net == NET_IPV4 || net == NET_IPV6; }
```
nit: `IsClearnet()` is more our usual style.

fjahr commented on src/net.h:
```diff
@@ -1398,6 +1402,9 @@ class CConnman
     bool MultipleManualOrFullOutboundConns(Network net) const EXCLUSIVE_LOCKS_REQUIRED(m_nodes_mutex);
 
+    /** Returns true if outbound IPv4/IPv6 connections must be v2 only. */
+    bool RequiresV2ForOutbound(const CNetAddr &addr) const;
```
Seems like this could be private?

fjahr commented on src/init.cpp:
```diff
@@ -1024,6 +1027,12 @@ bool AppInitParameterInteraction(const ArgsManager& args)
 
+    if (args.GetBoolArg("-v2onlyclearnet", DEFAULT_V2_ONLY_CLEARNET)) {
+        if (args.GetBoolArg("-listen", DEFAULT_LISTEN)) {
+            return InitError(_("Cannot set -v2onlyclearnet=1 with listen=0. Current listen=1 provides valuable network capacity. Use -v2onlyclearnet only if you truly need encrypted-only outbound traffic."));
```
This error message is a bit confusing, first "Cannot set -v2onlyclearnet=1 with listen=0" seems to be wrong, it should say "listen=1" there (or s/with/without/). And then maybe switch the order of the current second and third sentence, I think that would read a bit better. For my taste the second sentence is a bit much because I don't think it's going to change anyone's mind but I would be fine to keep it. But we could also ask people to read the help text of the option carefully where we are more detailed.

stratospher replied:
oh whoops, thank you for the great catch!

maybe also makes sense to keep the error message short + keep the context only in the help text - it also prevents duplication. I've changed the error message to : `"Cannot set -v2onlyclearnet=1 with listen=1. See -help for details on -v2onlyclearnet."` open to better wordings/suggestions as well or if people feel we should give slightly more context.

fjahr commented on src/init.cpp:
```diff
@@ -569,6 +569,7 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
     argsman.AddArg("-i2pacceptincoming", strprintf("Whether to accept inbound I2P connections (default: %i). Ignored if -i2psam is not set. Listening for inbound I2P connections is done through the SAM proxy, not by binding to a local address and port.", DEFAULT_I2P_ACCEPT_INCOMING), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-onlynet=<net>", "Make automatic outbound connections only to network <net> (" + Join(GetNetworkNames(), ", ") + "). Inbound and manual connections are not affected by this option. It can be specified multiple times to allow multiple networks.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-v2transport", strprintf("Support v2 transport (default: %u)", DEFAULT_V2_TRANSPORT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-v2onlyclearnet", strprintf("Ensure all outbound IPv4/IPv6 peers use encrypted network traffic (default: %u). Using this option requires -listen=0 and takes valuable listening capacity away from the network. Enable this option only if passive network observers like ISPs, firewalls, etc. pose a threat and unencrypted network traffic must be avoided. Note: Encryption protects me ...
```
Not a strong opinion but as it is worded now, probably everyone who runs a bitcoin node will say, yes, these network observers are a problem for my privacy. It might be better to say something like: "if you are at risk of your traffic getting blocked or service shut down". That would be a bit more strongly steer people away from this if they don't know what they are doing/don't really need it. But it's just an idea, feel free to ignore.

vasild replied:
> It might be better to say something like: "if you are at risk of your traffic getting blocked or service shut down"

If you are at such a risk, then `-v2onlyclearnet=1` will not help. Mentioning "passive" is good because `-v2onlyclearnet=1` protects only against "passive" observers and not against "active" ones who can MITM themselves and spy on the encrypted traffic. Then the user has no way to ensure that a possible adversary will stay "passive" and will not switch to "active" at any moment.

fjahr commented on src/net.cpp:
```diff
@@ -444,6 +444,9 @@ CNode* CConnman::ConnectNode(CAddress addrConnect,
 
     for (auto& target_addr : connect_to) {
+        if (RequiresV2ForOutbound(target_addr) && !use_v2transport) {
+            continue;
```
Seems like this filtering could already be done in the `resolved` loop above so that the addr doesn't even end up here?

fjahr replied:
nit: In the commit message here: `s/mechainm/mechanism/`

stratospher replied:
I like the idea but the resolved for loop only resolves a hostname to the possible IP addresses (ex: `thisnode.com` to `ipaddress1`, and maybe a backup `ipaddress2`). if we do move the check up, we'd also need to add it in 3 spots whenever we do `connect_to.push_back`. so haven't changed it.

fjahr commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -129,6 +131,27 @@ def run_test(self):
         assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v1")
         check_node_connections(node=node0, num_in=1, num_out=0)
 
+        conf = Socks5Configuration()
```
nit: In commit message `s/conneections/connections/`

fjahr reviewed (commented):
> However this feels like lots of users setting listen=0 problem, and we're showing someone a reason for setting listen=0 + making that future slightly more probable. Will need to think more about this and curious about other people's thoughts on https://github.com/bitcoin/bitcoin/pull/30951#pullrequestreview-3429992454 as well.

Re-reading the code and @mzumsande's argument, I think I am now slightly leaning towards not tying it to `listen=0` but rather only disallowing v1 connections. But I would still be fine with the current approach as well, especially if we can find a bit better wording.

stratospher commented:
thank you for the great catch @fjahr! I've addressed your comments in [this push](https://github.com/bitcoin/bitcoin/compare/1e61206583d87ff0bc0fd7e245cefce4203b423f..263c16b537e1073a45e829bb98506db3ef662698). I've made the InitError message shorter but kept the help text as it is.

also summarising my understanding of the 2 implementation options possible for `v2onlyclearnet` in case it's helpful to reviewers:

purpose is assurance that passive observers can't see your encrypted traffic content (it doesn't hides the fact that you're running a Bitcoin node nor protect you from active adversaries)


|                          | Option A: outbound v2 only + no inbound connections (current PR) | Option B: outbound + inbound v2 only connection |
|--------------------------|------------------------------------------------------------------|------------------------------------------------|
| **Flag interaction**     | there's a flag interaction `-v2onlyclearnet=1` requires `-listen=0`   | no flag interaction                            |
| **Outbound IPv4/IPv6**   | v2 only                                                 | v2 only                               |
| **Outbound Tor/I2P/CJDNS** | v1 and v2 (those networks encrypt anyway)            | v1 and v2 (those networks encrypt anyway) |
| **Inbound IPv4/IPv6**    | none accepted since `-listen=0`                                     | v2 only accepted                               |
| **Inbound Tor/I2P/CJDNS** | no Tor/CJDNS inbound but I2P inbounds possible when overriding default with `-i2pacceptincoming` since it doesn't use a listening socket     | v1 and v2 accepted (those networks encrypt anyway) |
| **Advantage**            | no v1-only eclipse risk                                          | preserves listening capacity on the network    |
|                          |  Requiring -v2onlyclearnet=1 with -listen=0 would be a more private way of running a clearnet node since we also enforce no (maybe spy) node being able to connect to us |  (in option B, it is upto the user to configure whether they want inbounds or not)                                             |
|                          | avoids easy detection that you're running a bitcoin node by stateless packet connection. (traffic analysis which is more effort will give you away)  | (in option B, listening nodes still advertise their IP on the network and is easier to discover compared to traffic analysis. traffic analysis which is more effort will give you away here as well.)|
| **Disadvantage (hypothetical scenario when most of network switch on the option)** | total listening capacity on the network shrinks | risk of eclipsing legacy v1-only nodes. if many nodes accept only v2 inbound, an attacker can cheaply spin up many v1-accepting nodes and dominate the v1 node's peer set. |


which implementation option is better?
1. hypothetical worst case scenario - option B wins here but don't think we should base our decision entirely on this. Let's say some day far in the future if there are <50 v1 nodes and a large fraction of the network turn on `v2onlyclearnet`
- the network not having enough listening capacity since `-v2onlyclearnet` also forces `-listen=0` (option A) is more dangerous than a v1 node not being able to connect to a v2 node and being eclipsed (option B).
- chance of the hypothetical worst case scenario where most of the network turns on this flag is also unlikely. so don't think we should base our decision entirely on this. we already have a long list of opt-in flags that would be bad if huge fraction of network enabled them.
2. user configurability - option A (or) B can win here
	- (option A) if user expects more to come along with the option - since passive adversaries can't read the message contents now - they might also expect passive adversary to not able to statelessly connect to them and detect their node.
	- (option B) if user likes to configure things themselves.
3. option A wins here - there will be 1 extra condition in version message processing logic in net_processing for option B. we let them connect and while processing version message discover it's V1 and then disconnect. don't think it's important but just noting it here.
4. option B wins here - more elegant in the sense that this option will naturally phase out and we can remove it in the far future when everyone is in v2 :)
5. option A wins here - we don't want passive adversary statelessly connecting and detecting node.

I personally am ok with both. Maybe slight inclination to B since it gives the user configuration knob to do what they want. And we could document it in the help text to mention using listen=0 as well. Not sure what others prefer though. 

Would love thoughts on what people prefer before changing approach/pushing an update :)

andrewtoth commented:
> a user who doesn't want their traffic spied on probably shouldn't accept arbitrary inbounds either

Can you elaborate on how v2 without inbounds makes traffic analysis more difficult than v2 with inbounds? I'm not a traffic analysis expert, and to me it would make sense that any analysis happening on 10 v2 connections vs 125 v2 connections would come to the same conclusion about whether or not you were running a bitcoin node? I'm fairly sure I'm missing something though... :)

But otherwise I would lean towards requiring `listen=0`, since it would make node runners who were interested in providing inbound connections hesitate to turn this option on. Without requiring that, it seems too easy for everyone to just turn it on without considering existing v1 nodes.

w0xlt commented on src/netbase.h:
```diff
@@ -179,6 +179,7 @@ enum Network ParseNetwork(const std::string& net);
 std::string GetNetworkName(enum Network net);
 /** Return a vector of publicly routable Network names; optionally append NET_UNROUTABLE. */
 std::vector<std::string> GetNetworkNames(bool append_unroutable = false);
+bool IsClearnet(enum Network net);
```
nit: A Doxygen comment can be added. 

```suggestion
/** Returns true if the network is IPv4 or IPv6 (as opposed to a privacy network like Tor/I2P/CJDNS). */
bool IsClearnet(enum Network net);
```

w0xlt commented on src/net.h:
```diff
@@ -77,6 +77,8 @@ static const int MAX_FEELER_CONNECTIONS = 1;
 /** -listen default */
 static const bool DEFAULT_LISTEN = true;
+/** -v2onlyclearnet default */
+static const bool DEFAULT_V2_ONLY_CLEARNET = false;
```
nit:

```suggestion
static constexpr bool DEFAULT_V2_ONLY_CLEARNET{false};
```

w0xlt commented on src/net.h:
```diff
@@ -1755,6 +1760,13 @@ class CConnman
+     * outbound connections on IPV4/IPV6 need to be v2 connections.
+     * outbound connections on Tor/I2P/CJDNS can be v1 or v2 connections.
+     */
+    bool m_v2only_clearnet;
```
If code ever reads `m_v2only_clearnet` before `Init()` is called, it's UB.

```suggestion
    bool m_v2only_clearnet{false};
```

w0xlt commented on src/init.cpp:
```diff
@@ -574,6 +569,7 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
     argsman.AddArg("-i2pacceptincoming", strprintf("Whether to accept inbound I2P connections (default: %i). Ignored if -i2psam is not set. Listening for inbound I2P connections is done through the SAM proxy, not by binding to a local address and port.", DEFAULT_I2P_ACCEPT_INCOMING), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-onlynet=<net>", "Make automatic outbound connections only to network <net> (" + Join(GetNetworkNames(), ", ") + "). Inbound and manual connections are not affected by this option. It can be specified multiple times to allow multiple networks.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-v2transport", strprintf("Support v2 transport (default: %u)", DEFAULT_V2_TRANSPORT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-v2onlyclearnet", strprintf("Ensure all outbound IPv4/IPv6 peers use encrypted network traffic (default: %u). Using this option requires -listen=0 and takes valuable listening capacity away from the network. Enable this option only if passive network observers like ISPs, firewalls, etc. pose a threat and unencrypted network traffic must be avoided. Note: Encryption protects me ...
```
```suggestion
    argsman.AddArg("-v2onlyclearnet", strprintf("Ensure all outbound IPv4/IPv6 peers use encrypted network traffic (default: %u). Using this option requires -listen=0 and takes valuable listening capacity away from the network. Enable this option only if passive network observers like ISPs, firewalls, etc. pose a threat and unencrypted network traffic must be avoided. Note: Encryption protects message contents but does not obscure that you are running a Bitcoin node. Peers can still connect to you, observers can still identify Bitcoin activity from your outbound connection attempts to the default port (8333) or by analysing traffic patterns.", DEFAULT_V2_ONLY_CLEARNET), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```

w0xlt commented on src/net.cpp:
```diff
@@ -444,6 +444,9 @@ CNode* CConnman::ConnectNode(CAddress addrConnect,
     std::unique_ptr<i2p::sam::Session> i2p_transient_session;
 
     for (auto& target_addr : connect_to) {
+        if (RequiresV2ForOutbound(target_addr) && !use_v2transport) {
```
If a user enables this option and wonders why connections to certain peers aren't happening, there's nothing in the debug log to explain it.

```suggestion
        if (RequiresV2ForOutbound(target_addr) && !use_v2transport) {
            LogDebug(BCLog::NET, "skipping v1 connection to %s (-v2onlyclearnet)\n", target_addr.ToStringAddrPort());
```

w0xlt commented on src/net.cpp:
```diff
@@ -444,6 +444,9 @@ CNode* CConnman::ConnectNode(CAddress addrConnect,
     std::unique_ptr<i2p::sam::Session> i2p_transient_session;
 
     for (auto& target_addr : connect_to) {
+        if (RequiresV2ForOutbound(target_addr) && !use_v2transport) {
```
When `pszDest` is unresolved because DNS is delegated to a name proxy, it might be better to treat it as possibly IPv4/IPv6 and block v1 under `-v2onlyclearnet`.

```suggestion
        const bool possibly_clearnet_name{pszDest && !target_addr.IsValid()};
        if ((RequiresV2ForOutbound(target_addr) ||
             (m_v2only_clearnet && possibly_clearnet_name)) &&
            !use_v2transport) {
```

P.S.: Same logic can be applied to `DisconnectNodes()`.

```diff
                 // and we don't want to hold up the socket handler thread for that long.
-                if (network_active && !RequiresV2ForOutbound(pnode->addr) && pnode->m_transport->ShouldReconnectV1()) {
+                const bool possibly_clearnet_name{
+                    m_v2only_clearnet && !pnode->addr.IsValid() && !pnode->m_dest.empty()};
+                if (network_active && !RequiresV2ForOutbound(pnode->addr) &&
+                    !possibly_clearnet_name && pnode->m_transport->ShouldReconnectV1()) {
                     reconnections_to_add.push_back({
                         .addr_connect = pnode->addr,
```

stratospher replied:
oh nice edge case! I've taken both your suggestions and placed it inside `RequiresV2ForOutbound`

davidgumberg commented:
> Can you elaborate on how v2 without inbounds makes traffic analysis more difficult than v2 with inbounds? I'm not a traffic analysis expert, and to me it would make sense that any analysis happening on 10 v2 connections vs 125 v2 connections would come to the same conclusion about whether or not you were running a bitcoin node?

I think there are two issues if you don't want to be discovered that you are running a node and have `listen=1` set: 

1. There are messages being passed around (on v1 connections for as long as there are v1 connections) with your address written on it that say "This address is a bitcoin node!".
2. Any hostile network can probe that you are running a node by attempting to v2 handshake you, and honestly this is not a large escalation in threat model, this is a widely deployed technique by many passive observers. (https://dl.acm.org/doi/epdf/10.1145/2815675.2815690) 

That being said, I still lean toward option B, of not tying `-v2only` to `listen=0`, even though in many cases one wants both, there are also cases that may want one without the other (you don't care about being detected as a node but you want to make it harder for a passive observer to learn anything interesting from your bitcoin traffic). Maybe in the future it would be good to have a single flag that enables sane defaults for privacy or some privacy document describing the shortcomings and tradeoffs of the available options.

I am Concept +0 leaning towards Concept ACK on this, I agree that given all of the other privacy issues that exist in both Bitcoin P2P and Bitcoin Core's implementation there won't be a step change in privacy from doing this; but I am very skeptical of this species of argument since the same arguments are used circularly to persist the other practices that degrade privacy persist, e.g. port 8333 (why change the port if we are detectable anyways via v1 connections?) Yes there are other problems, and they ought to be addressed, but raising the cost seems a good direction.

The only concern I feel hasn't been clearly addressed is the depletion of inbound v1 slots on the network; but maybe this isn't an issue needing discussing yet since so few will run with this option enabled.

stratospher commented:
pushed an [update](https://github.com/bitcoin/bitcoin/compare/263c16b537e1073a45e829bb98506db3ef662698..b559f2fdb1adf6bddc467751d6da1fc8e02b78ee ) addressing @w0xlt's comments. thanks @w0xlt  for the great catches! And thanks @andrewtoth and @davidgumberg  for sharing your opinions!

The PR still implements [option A](https://github.com/bitcoin/bitcoin/pull/30951#issuecomment-4346061700) (node with `-v2onlyclearnet=1` option does outbound only connection)

> Can you elaborate on how v2 without inbounds makes traffic analysis more difficult than v2 with inbounds? I'm not a traffic analysis expert, and to me it would make sense that any analysis happening on 10 v2 connections vs 125 v2 connections would come to the same conclusion about whether or not you were running a bitcoin node? I'm fairly sure I'm missing something though... :)

@andrewtoth you're right that they're all the same from a traffic analyser PoV. I should reword/remove that sentence. what I meant was - if we couple `-v2onlyclearnet=1` with `-listen=0`, we can get both:
1. protection against passive observer who can see plaintext traffic.
2. some protection against active adversary/probers who can connect to you and see and analyse your traffic

It was an assumption that users who care about (1) might often care about (2). Requiring `-v2onlyclearnet=1` with `-listen=0` would be a more private way of running a clearnet node. 

Though I personally prefer letting operators choose what they want to run: (1) becomes the responsibility of `-v2onlyclearnet`, (2) becomes the responsibility of `-listen`.


> The only concern I feel hasn't been clearly addressed is the depletion of inbound v1 slots on the network; but maybe this isn't an issue needing discussing yet since so few will run with this option enabled.

@davidgumberg I do think it is important as well since this is the main blocker concern for doing option B. Also not sure how to reason about it since it's hypothetical. It is very unlikely that all v2 listening nodes turn on `-v2onlyclearnet`. And yes in case majority of the v2 nodes turn on this option - v1 nodes won't find nodes to connect to and are at a risk. But how realistic is it that everyone turns on this option. It worked for `onlynet` and `listen` in the past though that doesn't necessarily guarantee it will work for us now.

> But otherwise I would lean towards requiring listen=0, since it would make node runners who were interested in providing inbound connections hesitate to turn this option on. Without requiring that, it seems too easy for everyone to just turn it on without considering existing v1 nodes.

Right. Good point about the hesitation. I could imagine operators turning this on to prevent the v1 node data crawlers/probers that exist today and leach on the inbound slots. Once those crawlers/probers add v2 handshake support (which they should easily be able to nowadays), that motivation disappears as well.

Maybe a warning might make the node operator reconsider if they really need it?

```
  if (args.GetBoolArg("-v2onlyclearnet", DEFAULT_V2_ONLY_CLEARNET) &&
      args.GetBoolArg("-listen", DEFAULT_LISTEN)) {                                                                                                                         
      InitWarning(_("-v2onlyclearnet=1 is set with listening enabled: v1 inbound "                                                                                          
                    "connections on IPv4/IPv6 will be rejected. If many nodes adopt "                                                                                       
                    "this configuration, older v1-only nodes may struggle to find "                                                                                         
                    "inbound peers. Do you really need it?"));                                                                                                                                     
  }
```

naiyoma commented:
this is  a good improvement. While it doesn’t fully solve the traffic analysis issue, making it harder is still a positive in my view.
I also like that we’re explicitly communicating that it’s still possible for an observer to tell when someone is running a node .
I’m leaning toward option A here, mainly because it feels like the safest option.
But I  a bit hesitant about the tradeoff. I think it’s better to think about this under the assumption that most nodes will adopt this option. In that case, an eclipse attack on v1 seems like a much worse outcome than the impact of listening traffic shrinking.

naiyoma commented on src/net.h:
```diff
@@ -1774,6 +1772,13 @@ class CConnman
+     * outbound connections on IPV4/IPV6 need to be v2 connections.
+     * outbound connections on Tor/I2P/CJDNS can be v1 or v2 connections.
+     */
+    bool m_v2only_clearnet{false};
```
micro nit: in 760354a0397859430bed2f861289de6f591aafab
``` diff
-    bool m_v2only_clearnet{false};
+    bool m_v2only_clearnet{DEFAULT_V2_ONLY_CLEARNET};
```

stratospher replied:
done. thanks!

danielabrozzoni commented on src/netbase.h:
```diff
@@ -180,6 +179,8 @@ enum Network ParseNetwork(const std::string& net);
 std::string GetNetworkName(enum Network net);
 /** Return a vector of publicly routable Network names; optionally append NET_UNROUTABLE. */
 std::vector<std::string> GetNetworkNames(bool append_unroutable = false);
+/** Returns true if the network is IPv4 or IPv6 (as opposed to Tor/I2P/CJDNS). */
```
nit: In 0f32139ea080074f4d73b0e63747bae0fa8c74b3 (comment in a random place):
I think you should also use `IsClearnet` in:
- https://github.com/bitcoin/bitcoin/blob/ea9afb61a1c52979d0e64e4dae9b3a611f7b9a5e/src/net.cpp#L3149
- https://github.com/bitcoin/bitcoin/blob/ea9afb61a1c52979d0e64e4dae9b3a611f7b9a5e/src/init.cpp#L865

danielabrozzoni commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -129,6 +131,29 @@ def run_test(self):
+        self.restart_node(0, extra_args=args)
+        self.log.info("Test -v2onlyclearnet=1 behaviour")
+        self.log.info("Check that outbound v2 connection to an ipv4 peer is successful")
+        node0.addnode("15.61.23.23:1234", "onetry", True)
```
nit (here and below): can you pass `v2transport=` explicitly? 
I found it a bit hard to see where the test was selecting v1 vs v2 transport :)

```diff
diff --git a/test/functional/p2p_v2_encrypted.py b/test/functional/p2p_v2_encrypted.py
index e2acaf71a0..03e12e306a 100755
--- a/test/functional/p2p_v2_encrypted.py
+++ b/test/functional/p2p_v2_encrypted.py
@@ -142,11 +142,15 @@ class P2PEncrypted(BitcoinTestFramework):
         self.restart_node(0, extra_args=args)
         self.log.info("Test -v2onlyclearnet=1 behaviour")
         self.log.info("Check that outbound v2 connection to an ipv4 peer is successful")
-        node0.addnode("15.61.23.23:1234", "onetry", True)
+        node0.addnode("15.61.23.23:1234", "onetry", v2transport=True)
         assert_equal(node0.getpeerinfo()[-1]["addr"], "15.61.23.23:1234")
         self.log.info("Check that outbound v1 connection to an ipv4 peer is unsuccessful")
         with node0.assert_debug_log(expected_msgs=["skipping v1 connection to 8.8.8.8:1234"]):
-            node0.addnode("8.8.8.8:1234", "onetry", False)
+            node0.addnode("8.8.8.8:1234", "onetry", v2transport=False)
```

danielabrozzoni reviewed (commented):
Concept ACK!

Regarding whether to allow incoming connections, I’m slightly leaning towards disallowing them.
(Thanks for the summary table @stratospher, it's very useful!)

In the hypothetical worst-case scenario, option A would be worse because the network might not have enough inbound capacity. However I agree with @andrewtoth:

> But otherwise I would lean towards requiring listen=0, since it would make node runners who were interested in providing inbound connections hesitate to turn this option on. Without requiring that, it seems too easy for everyone to just turn it on without considering existing v1 nodes.

For this reason, I think disallowing incoming connections is better, as it lowers the chance of the worst-case scenario.

---


I have a question: throughout this thread, we talk about how with `-v2onlyclearnet` a passive observer can't identify a Bitcoin node by looking for network magic bytes in the traffic - although there's other relatively easy ways to spot one, such as looking for connections from/to port 8333.

If we implemented option B, would this still hold? I imagine the implementation would be: if a node connects to us using v1, we close the connection. I think that would mean that a passive observer could still observe the network magic. Does this make sense? 

---

I also have some implementation questions.

First of all, I'm wondering if we should disallow incoming, or disallow incoming *on clearnet* - since these are the connections that are unencrypted. (I'm not sure this can be easily done with the current startup flags)

I think we currently disallow incoming Tor connections because `-listen=0` sets `-listenonion=0`, but incoming I2P connections are still allowed if explicitly requested. This works:

```diff
diff --git a/test/functional/feature_config_args.py b/test/functional/feature_config_args.py
index bdb5c5293e..8e1478e868 100755
--- a/test/functional/feature_config_args.py
+++ b/test/functional/feature_config_args.py
@@ -442,6 +442,15 @@ class ConfArgsTest(BitcoinTestFramework):
         self.nodes[0].assert_start_raises_init_error(
             extra_args=['-v2onlyclearnet=1', '-v2transport=1', '-listen=1'],
             expected_msg='Error: Cannot set -v2onlyclearnet=1 with listen=1. See -help for details on -v2onlyclearnet.')
+        args = [
+            "-v2onlyclearnet=1",
+            "-v2transport=1",
+            "-listen=0",
+            "-i2psam=127.0.0.1:3333",
+            "-nobind",
+            "-i2pacceptincoming=1",
+        ]
+        self.restart_node(0, extra_args=args)
 
```

Also, we currently allow v1 connections to not publicly routable addresses, so you can addnode to a local network address, and connect to it through v1. Is this okay? I guess it doesn't pose a privacy concern in most cases, but maybe it makes sense to fix for consistency? This currently works:

```diff
diff --git a/test/functional/p2p_v2_encrypted.py b/test/functional/p2p_v2_encrypted.py
index e2acaf71a0..880ec70972 100755
--- a/test/functional/p2p_v2_encrypted.py
+++ b/test/functional/p2p_v2_encrypted.py
@@ -147,6 +147,8 @@ class P2PEncrypted(BitcoinTestFramework):
         self.log.info("Check that outbound v1 connection to an ipv4 peer is unsuccessful")
         with node0.assert_debug_log(expected_msgs=["skipping v1 connection to 8.8.8.8:1234"]):
             node0.addnode("8.8.8.8:1234", "onetry", False)
+        with node0.assert_debug_log(expected_msgs=[], unexpected_msgs=["skipping v1 connection to 192.168.0.1:1234"]):
+            node0.addnode("192.168.0.1:1234", "onetry", v2transport=False)
         assert all(peer["addr"] != "8.8.8.8:1234" for peer in node0.getpeerinfo())
         self.log.info("Check that outbound v1 connection to an onion peer is successful")
         addr = "pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion:8333"
```

(You can see in logs that it doesn't really connect to 192.168.0.1, but it attemps the connection despite being v1)

stratospher committed:
doc: add release notes for v2onlyclearnet option

stratospher commented:
thank you for the thorough review @danielabrozzoni! I've addressed your review comments in [this push](https://github.com/bitcoin/bitcoin/compare/b559f2fdb1adf6bddc467751d6da1fc8e02b78ee..d72df0fc8369efc391e3f777e1b9fb20a223beb3) and added a [commit](https://github.com/bitcoin/bitcoin/pull/30951/changes/d72df0fc8369efc391e3f777e1b9fb20a223beb3) for release notes.

> I think that would mean that a passive observer could still observe the network magic. Does this make sense?

yes if we allow inbound - they'd see the network magic and a few more bytes of the first message before we drop the connection. rest of the v2 network traffic (if not dropped) would be encrypted.

> I think we currently disallow incoming Tor connections because -listen=0 sets -listenonion=0, but incoming I2P connections are still allowed if explicitly requested.

ooh good find! I've corrected the [summary table](https://github.com/bitcoin/bitcoin/pull/30951#issuecomment-4346061700).

ideally we'd drop only the unencrypted (clearnet) inbounds and keep the Tor/I2P/CJDNS ones, since those are already encrypted. I think it will be awkward to implement though - I2P unlike the others doesn't use a listening socket which is why we can override it.

there is a soft-set in the init.cpp where `-listen=0` -> `-listenonion=0` and `-i2pacceptincoming=0`. I feel it's ok if the node operator really wants to override it to allow i2p inbound since the option's guarantees still hold and there's no privacy leak. Maybe we should just document this behaviour somewhere? I tried putting it in the release notes but felt it read too confusing for end users. 

> we currently allow v1 connections to not publicly routable addresses, so you can addnode to a local network address, and connect to it through v1. 

good point. I think it's fine to leave as is - I've added a comment in `RequiresV2ForOutbound` rather than forcing v2 for LAN.

gmaxwell commented:
An incremental option to address partitioning concerns is to only allow inbound v1 connections that are fRelay=false.  While this doesn't achieve all the benefits of v2only, including lowering the visibility of plaintext traffic (including potential plaintext malware that trips up network scanning software), many concerns favoring v2 are related specifically to transaction relay.  As an added bonus the dnsseed indexing run today doesn't use v2, but it does connect with fRelay=false.

I cannot express how strongly I disagree with forcing listen=0.  It is is an outright anti-feature to unnecessarily tie a security/privacy setting to providing resources to others-- it's analogous to saying that anyone running a limited node (prune>0) must be listen=0 or anyone not offering blockfilters must be listen=0.

There is no serious risk of abandoning older node through an off-by-default configuration setting. And failing to provide network capacity means that there are less sockets available for v2 connections which means other people's connections will use up v1 accommodating sockets that would otherwise be available-- so that move would have precisely opposite the intended effect purely by lowering the total number of sockets available in the network.  (To whatever extent a manually configured option will have any effect at all).

It isn't like users get paid or something for accepting listening connections as some incentive to accept connections from older peers, -- and to whatever extent users are hurt by not listening  (e.g. by getting a reduced txn relay anonymity set) then that is an argument to not have the restriction.   Moreover, in terms of transaction privacy privacy against a global passive observer is greatly improved to the extent that a users broadcast crosses more of the network through peers without attacker observed connections (so encryption is a pre-req), but v2only nodes can hardly contribute in this way if they don't accept connections.  I might go so far as to say that a v2 only listen=0 peer is pretty pointless, might as well just run a non-listening node that connects out exclusively through tor and forget about the v1/v2 distinction.

The question of leaving v1 out in the cold would be very material if considering making this a default option, but that isn't under discussion.  That said, Satoshi had no problem changing the p2p protocol in a way that was only backwards compatible for two years.  I don't see any great reason to run into disabling v1 by default today, but there may be one in the future (such as the further spread of network based anti-malware that can by triggered by traffic in the p2p connection). If there were it would be important that the existing v2 only nodes weren't listen=0.

A separate issue that I don't see discussed or addressed in the commits:  Addresses learned via dnsseed right now as well as the static addr databases don't have the v2 flag.  And so a node that has v2 as a requirement for its peer selection won't find any peers at all from a cold start and will just sit connected to nothing. I think to address that, the seeds database should just be filtered to v2 peers (as there are plenty and already outdated peers get excluded) and treated as v2 in addrman,  and similar dnsseeds should provide a name to filter on that service bit as was done for NODE_WITNESS.

mzumsande commented:
> I cannot express how strongly I disagree with forcing listen=0.

FWIW, I also still prefer not doing that, https://github.com/bitcoin/bitcoin/pull/30951#pullrequestreview-3429992454

> I think to address that, the seeds database should just be filtered to v2 peers (as there are plenty and already outdated peers get excluded) and treated as v2 in addrman, and similar dnsseeds should provide a name to filter on that service bit as was done for NODE_WITNESS.

Good point, but maybe it's not strictly necessary to change the dns seeds, we could also just switch the default to v2 for all results from dns seeds and fixed seeds - if it happens to be wrong, a regular node would just downgrade to v1. And given that most nodes support v2 by now, it's very unlikely that a v2-only node would only receive v1 peers from dns seeds.

gmaxwell commented:
Oh good point on the seeds. I agree, that's tidy enough.

mzumsande commented:
proposed in #35766, I think it makes sense to change regardless of this PR.

sedited commented:
@stratospher can you rebase to get the changes from #35766 in here?

stratospher commented:
(just a rebase. will address the comments here soon!)

fanquake commented:
> (just a rebase. will address the comments here soon!)

@stratospher note that this needs rebase again, and feature freeze is in 2 days (this is tagged for v32.0).

mzumsande commented:
looks to me that there is no agreement among reviewers about coupling it to `listen=0`, so unless that changes very quickly, maybe we have to postpone for v33?

sedited commented:
Yes, postponing this.
